### ISSUE TITLE
Allow to use regular expressions in whitelist

### ISSUE THREAD

**Issue by Ana06** [bkuhlmann/git-cop#9]
Allow to use regular expressions in whitelist qould be really cool! I would specially find it useful in the `commit_subject_prefix`, as you could for example allow any word ending with `ed`, enforce that the first letter of the commit message is capitalized or support tags as discussed in https://github.com/bkuhlmann/git-cop/issues/8. Although I don't see any reason to not make this more generic and allow also this in the whitelists and blacklists of other cops.

**Comment by bkuhlmann** [comment]
Yep, I can see this being useful. ...and, to be honest, I had considered it as a solution in the original implementation. I agree, if this were to be supported, we should make it consistent behavior across all lists. There are some pros and cons to this approach that you might want to consider:

**Pros:**

- More flexibility in what is flagged as an error, warning, etc.

**Cons:**

- Regular expressions, when written incorrectly (and this is easy to do), can match too much or very little. This can yield false positives.
- Hard to test. Not so much from a gem maintainers point of view but from the user of the gem. For instance, this gem currently supports explicit whitelist/blacklists and provides sufficient test coverage that verifies the lists behave as designed. If there are edge cases or bugs, then they are easier to account for and nail down so that a user of this gem has a stronger guarantee that this gem has sufficient test coverage *and* is doing more work on behalf of the user. With regular expressions, this gem can have test coverage for them too but it can no longer guarantee well defined behavior (well, to a lesser extent at least). That means any regular expression has to be well tested by user of this gem. In other words, more onus is put on the user of the gem rather than the gem itself. Not saying this is a bad situation but there is something to be said about making a tool that is easy to use and has a higher guarantee of predictability than something that is too easily configurable and allows the user to shoot themselves in the foot per se.

Out of curiosity, what does the complete tag list (all combinations) of what you're trying to account for look like? Would it be OK to publish that list here or is that private information only? I'd be curious to see what that looks like. If the list consists of an exponential combination of tags, then you don't need to assemble it. I'm curious, is all.

**Comment by Ana06** [comment]
> Regular expressions, when written incorrectly (and this is easy to do), can match too much or very little. This can yield false positives.

I guess normally the regular expression won't be that complicate. And if it is, it is the responsibility of the person who wrote it to ensure that it is right. :smile: 

> Hard to test. Not so much from a gem maintainers point of view but from the user of the gem. For instance, this gem currently supports explicit whitelist/blacklists and provides sufficient test coverage that verifies the lists behave as designed. If there are edge cases or bugs, then they are easier to account for and nail down so that a user of this gem has a stronger guarantee that this gem has sufficient test coverage and is doing more work on behalf of the user. With regular expressions, this gem can have test coverage for them too but it can no longer guarantee well defined behavior (well, to a lesser extent at least). That means any regular expression has to be well tested by user of this gem. In other words, more onus is put on the user of the gem rather than the gem itself. Not saying this is a bad situation but there is something to be said about making a tool that is easy to use and has a higher guarantee of predictability than something that is too easily configurable and allows the user to shoot themselves in the foot per se.

Do you know [Rantly](https://github.com/rantly-rb/rantly) gem? It allows you to write property tests in RSpec and I found it really cool for testing regular expressions. Just because of all the problem you mention :wink:

> Out of curiosity, what does the complete tag list (all combinations) of what you're trying to account for look like? Would it be OK to publish that list here or is that private information only? I'd be curious to see what that looks like. If the list consists of an exponential combination of tags, then you don't need to assemble it. I'm curious, is all.

Well actually the project I am working on is free software, so there is nothing private: https://github.com/openSUSE/open-build-service/commits/master Although not all the commits there are perfect, the idea is basically introducing git-cop to help us to write better commits. :wink: The regular expression should be something like:

`/(\[backend\]|\[webui\]|\[api\]|\[ci\])+ (Fix|Add|Update|Remove|Refactor|Change|Enable|Disable|Test|Split)+ /`

**Comment by bkuhlmann** [comment]
> I guess normally the regular expression won't be that complicate. And if it is, it is the responsibility of the person who wrote it to ensure that it is right.

True, but it is always surprising to find those that log issues when it was their own doing. :)

> Do you know Rantly gem?

No, but thanks. That's a good tip!

> https://github.com/openSUSE/open-build-service/commits/master

Oh, my, you use a merge workflow instead of a rebase workflow. 😱  Yeah, I can see how this gem will help some. :)

An American holiday is happening this weekend, so I'm not sure how respondent I'll be. I hope to have some time to work on this but if you beat me to it, add yourself as an owner of this issue. Some quick requirements for implementing this:

- Needs to be able to accept strings or regular expressions for the whitelists/blacklists.
- The regular expressions should be anchored, if possible. (i.e. beginning of string for prefixes, end of string for suffixes).
- Can regular expressions be loaded from YAML configurations? I forget, but need to double check.

**Comment by Ana06** [comment]
> Needs to be able to accept strings or regular expressions for the whitelists/blacklists.

Do you mean that the current behavior of using string in the whitelist keep working?

> The regular expressions should be anchored, if possible. (i.e. beginning of string for prefixes, end of string for suffixes).

I am not sure what you mean with this :confused: 

> Can regular expressions be loaded from YAML configurations? I forget, but need to double check.

We can always load the reg exp as string and then parse it, right?

**Comment by bkuhlmann** [comment]
> Do you mean that the current behavior of using string in the whitelist keep working?

Yes, sorry, I wrote my original notes in haste but, yes, you are correct.

Regarding regular expressions...if someone supplies a custom regex in a white/black list, the implementation should anchor the regex. For example, let's say the Git commit subject prefix whitelist looks like this:

```
[
  /Added.+/,
  /Removed.+/
]
```

The implementation should add `\A` before the regex. The dynamically updated list looks more like this:

```
[
  /\AAdded.+/,
  /\ARemoved.+/
]
```

This ensures someone can't accidentally abuse the commit subject prefix cop by using a regex that looks for end of line. They still could but at least it would always be anchored at the beginning of the line.

> We can always load the reg exp as string and then parse it, right?

Yep, I think that should work.

**Comment by bkuhlmann** [comment]
This work has been resolved this [pull request](https://github.com/bkuhlmann/git-cop/pull/18).
